server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -4046,6 +4048,68 @@ protected void commitIndexWriter(final IndexWriter writer, final Translog transl
         }
     }
 
+    public void testFlushOnIdleAfterOp() throws Exception {
+        // Holding the write lock makes the index/delete op to halt before being processed by the engine
+        final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
+        final ReleasableLock readLock = new ReleasableLock(rwl.readLock());
+        final ReleasableLock writeLock = new ReleasableLock(rwl.writeLock());
+        IndexShard shard = newStartedShard(true, Settings.EMPTY, new IndexingOperationListener() {
+            @Override
+            public Engine.Index preIndex(ShardId shardId, Engine.Index operation) {
+                try (ReleasableLock lock = readLock.acquire()) {
+                    return operation;
+                }
+            }
+
+            @Override
+            public Engine.Delete preDelete(ShardId shardId, Engine.Delete delete) {
+                try (ReleasableLock lock = readLock.acquire()) {
+                    return delete;
+                }
+            }
+        });
+
+        indexDoc(shard, "_doc", "0");
+        indexDoc(shard, "_doc", "1");
+
+        // Do a flush on idle
+        long flushesBefore = shard.flushStats().getPeriodic();
+        shard.flushOnIdle(0);
+        assertBusy(() -> assertThat(shard.flushStats().getPeriodic(), equalTo(flushesBefore + 1)));
+        assertFalse(shard.isActive());
+
+        // Index or delete a doc and halt it before being processed by the engine
+        boolean indexElseDelete = randomBoolean();
+        Thread t = new Thread(() -> {
+            try {
+                if (indexElseDelete) {
+                    indexDoc(shard, "_doc", "2");
+                } else {
+                    deleteDoc(shard, "0");
+                }
+            } catch (IOException e) {
+                throw new AssertionError("failed while processing op [" + e.getMessage() + "]");
+            }
+        });
+        try (ReleasableLock lock = writeLock.acquire()) {
+            t.start();
+            assertBusy(() -> assertThat(rwl.getQueueLength(), equalTo(1)));
+            assertFalse(shard.isActive());
+        } // Allow op to complete
```

Comment:
Well, the race was happening if the flush on idle completed before the operation was fully processed. Now that we set the active flag after the operation has been processed, we do not have the race anymore.

Putting a flush on idle here will basically be a no-op because the `active` flag is already false and it will be skipped (the flush on idle actually flushes if the `active` flag is true).

Reply:
Right, a flush on idle will now be a noop so my suggestion won't make much sense.

Thanks for confirming my understanding